File: order-sequence-drift-after-migration/python/detect_sequence_drift.py
"""Flag Magento 2 order sequence drift after a data migration or DB restore.

Magento 2 numbers orders from dedicated sequence_order_<store> tables, tracked
via sales_sequence_meta and sales_sequence_profile, completely separate from
the sales_order table's own entity_id auto increment column. A migration from
Magento 1, or a manual DB import or restore, commonly copies sales_order rows
without correctly re-seeding the sequence table's last issued value, so the
next order minted from it can collide with an existing increment_id or skip a
huge range. There is no REST endpoint to rewrite sequence state, so this only
reports the drift and the recommended AUTO_INCREMENT reset value. Run on a
schedule. Safe to run again and again.
"""
import os
import sys
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("detect_sequence_drift")

MAGENTO_URL = os.environ["MAGENTO_URL"].rstrip("/")
TOKEN = os.environ["MAGENTO_ADMIN_TOKEN"]
GAP_THRESHOLD = int(os.environ.get("GAP_THRESHOLD", "1000"))
PAGE_SIZE = int(os.environ.get("PAGE_SIZE", "200"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"


def magento_get(path, params=None):
    r = requests.get(
        f"{MAGENTO_URL}/rest/V1{path}",
        params=params or {},
        headers={"Authorization": f"Bearer {TOKEN}"},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def orders_sorted_by_increment(page_size=200, current_page=1):
    params = {
        "searchCriteria[sortOrders][0][field]": "increment_id",
        "searchCriteria[sortOrders][0][direction]": "ASC",
        "searchCriteria[pageSize]": page_size,
        "searchCriteria[currentPage]": current_page,
    }
    return magento_get("/orders", params)["items"]


def all_orders_sorted_by_increment(page_size=200):
    page = 1
    while True:
        items = orders_sorted_by_increment(page_size, page)
        if not items:
            return
        for item in items:
            yield item
        if len(items) < page_size:
            return
        page += 1


def normalize_order(item):
    return {
        "entityId": item.get("entity_id"),
        "storeId": item.get("store_id"),
        "incrementId": item.get("increment_id"),
        "createdAt": item.get("created_at"),
    }


def strip_prefix(increment_id, prefix):
    value = increment_id[len(prefix):] if prefix and increment_id.startswith(prefix) else increment_id
    return int(value.lstrip("0") or "0")


def detect_sequence_drift(orders, prefix_by_store, gap_threshold=1000):
    """Pure decision function. No DB/HTTP calls, fully testable on synthetic
    order arrays.

    orders: list of {entityId, storeId, incrementId, createdAt}
    prefix_by_store: {storeId: prefix} used to strip a known prefix/suffix
        from incrementId before parsing it as a number.

    Returns {duplicates, gaps, maxNumericByStore}:
      - duplicates: same numeric value, more than one distinct entityId
      - gaps: consecutive numeric deltas beyond gap_threshold
      - maxNumericByStore: per-store max numeric value, the recommended
        next AUTO_INCREMENT seed is max + 1
    """
    by_store = {}
    for o in orders:
        by_store.setdefault(o["storeId"], []).append(o)

    duplicates = []
    gaps = []
    max_numeric_by_store = {}

    for store_id, store_orders in by_store.items():
        prefix = prefix_by_store.get(store_id, "")
        rows = sorted(
            (
                {
                    "entityId": o["entityId"],
                    "numeric": strip_prefix(o["incrementId"], prefix),
                    "incrementId": o["incrementId"],
                }
                for o in store_orders
            ),
            key=lambda r: r["numeric"],
        )

        seen = {}
        for r in rows:
            seen.setdefault(r["numeric"], []).append(r["entityId"])
        for numeric, entity_ids in seen.items():
            distinct = sorted(set(entity_ids))
            if len(distinct) > 1:
                duplicates.append({
                    "storeId": store_id,
                    "incrementId": next(r["incrementId"] for r in rows if r["numeric"] == numeric),
                    "entityIds": distinct,
                })

        for prev, curr in zip(rows, rows[1:]):
            gap_size = curr["numeric"] - prev["numeric"]
            if gap_size > gap_threshold:
                gaps.append({
                    "storeId": store_id,
                    "fromIncrement": prev["numeric"],
                    "toIncrement": curr["numeric"],
                    "gapSize": gap_size,
                })

        max_numeric_by_store[store_id] = max((r["numeric"] for r in rows), default=0)

    return {
        "duplicates": duplicates,
        "gaps": gaps,
        "maxNumericByStore": max_numeric_by_store,
    }


def run():
    raw_orders = list(all_orders_sorted_by_increment(PAGE_SIZE))
    orders = [normalize_order(item) for item in raw_orders]

    # No REST field exposes a store's increment_id prefix, so an empty prefix
    # is assumed unless overridden per store via PREFIX_BY_STORE, e.g. "1:ORD-,2:EU-".
    prefix_by_store = {}
    for pair in os.environ.get("PREFIX_BY_STORE", "").split(","):
        if ":" in pair:
            store_id, prefix = pair.split(":", 1)
            prefix_by_store[int(store_id.strip())] = prefix.strip()

    result = detect_sequence_drift(orders, prefix_by_store, GAP_THRESHOLD)

    for d in result["duplicates"]:
        log.warning(
            "Store %s: increment_id %s is duplicated across entity_id %s.",
            d["storeId"], d["incrementId"], d["entityIds"],
        )
    for g in result["gaps"]:
        log.warning(
            "Store %s: gap of %s between increment %s and %s.",
            g["storeId"], g["gapSize"], g["fromIncrement"], g["toIncrement"],
        )

    affected_stores = {d["storeId"] for d in result["duplicates"]} | {g["storeId"] for g in result["gaps"]}
    if affected_stores:
        for store_id in sorted(affected_stores, key=str):
            reset_value = result["maxNumericByStore"].get(store_id, 0) + 1
            log.error(
                "Store %s sequence drift detected. Recommended repair: "
                "ALTER TABLE sequence_order_%s AUTO_INCREMENT = %d (run by a DBA, not this script).",
                store_id, store_id, reset_value,
            )
        log.error("%d store(s) affected. Exiting non-zero. No sequence table was written.", len(affected_stores))
        sys.exit(1)
    else:
        log.info("Done. No sequence drift found across %d order(s).", len(orders))


if __name__ == "__main__":
    run()

File: order-sequence-drift-after-migration/node/detect-sequence-drift.js
/**
 * Flag Magento 2 order sequence drift after a data migration or DB restore.
 *
 * Magento 2 numbers orders from dedicated sequence_order_<store> tables, tracked
 * via sales_sequence_meta and sales_sequence_profile, completely separate from
 * the sales_order table's own entity_id auto increment column. A migration from
 * Magento 1, or a manual DB import or restore, commonly copies sales_order rows
 * without correctly re-seeding the sequence table's last issued value, so the
 * next order minted from it can collide with an existing increment_id or skip a
 * huge range. There is no REST endpoint to rewrite sequence state, so this only
 * reports the drift and the recommended AUTO_INCREMENT reset value. Run on a
 * schedule. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/magento/order-sequence-drift-after-migration/
 */
import { pathToFileURL } from "node:url";

const MAGENTO_URL = (process.env.MAGENTO_URL || "https://demo.example.com").replace(/\/$/, "");
const TOKEN = process.env.MAGENTO_ADMIN_TOKEN || "token_dummy";
const GAP_THRESHOLD = Number(process.env.GAP_THRESHOLD || 1000);
const PAGE_SIZE = Number(process.env.PAGE_SIZE || 200);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

export function stripPrefix(incrementId, prefix) {
  const value = prefix && incrementId.startsWith(prefix) ? incrementId.slice(prefix.length) : incrementId;
  const stripped = value.replace(/^0+/, "");
  return stripped === "" ? 0 : parseInt(stripped, 10);
}

/**
 * Pure decision function. No DB/HTTP calls, fully testable on synthetic
 * order arrays.
 *
 * orders: Array<{entityId, storeId, incrementId, createdAt}>
 * prefixByStore: Record<storeId, prefix> used to strip a known prefix/suffix
 *   from incrementId before parsing it as a number.
 *
 * Returns {duplicates, gaps, maxNumericByStore}:
 *   - duplicates: same numeric value, more than one distinct entityId
 *   - gaps: consecutive numeric deltas beyond gapThreshold
 *   - maxNumericByStore: per-store max numeric value, the recommended next
 *     AUTO_INCREMENT seed is max + 1
 */
export function detectSequenceDrift(orders, prefixByStore, gapThreshold = 1000) {
  const byStore = new Map();
  for (const o of orders) {
    if (!byStore.has(o.storeId)) byStore.set(o.storeId, []);
    byStore.get(o.storeId).push(o);
  }

  const duplicates = [];
  const gaps = [];
  const maxNumericByStore = {};

  for (const [storeId, storeOrders] of byStore) {
    const prefix = prefixByStore[storeId] || "";
    const rows = storeOrders
      .map((o) => ({
        entityId: o.entityId,
        numeric: stripPrefix(o.incrementId, prefix),
        incrementId: o.incrementId,
      }))
      .sort((a, b) => a.numeric - b.numeric);

    const seen = new Map();
    for (const r of rows) {
      if (!seen.has(r.numeric)) seen.set(r.numeric, []);
      seen.get(r.numeric).push(r.entityId);
    }
    for (const [numeric, entityIds] of seen) {
      const distinct = [...new Set(entityIds)].sort((a, b) => a - b);
      if (distinct.length > 1) {
        const match = rows.find((r) => r.numeric === numeric);
        duplicates.push({ storeId, incrementId: match.incrementId, entityIds: distinct });
      }
    }

    for (let i = 0; i < rows.length - 1; i++) {
      const gapSize = rows[i + 1].numeric - rows[i].numeric;
      if (gapSize > gapThreshold) {
        gaps.push({ storeId, fromIncrement: rows[i].numeric, toIncrement: rows[i + 1].numeric, gapSize });
      }
    }

    maxNumericByStore[storeId] = rows.length ? Math.max(...rows.map((r) => r.numeric)) : 0;
  }

  return { duplicates, gaps, maxNumericByStore };
}

async function magentoGet(path, params = {}) {
  const url = new URL(`${MAGENTO_URL}/rest/V1${path}`);
  for (const [k, v] of Object.entries(params)) url.searchParams.set(k, v);
  const res = await fetch(url, { headers: { Authorization: `Bearer ${TOKEN}` } });
  if (!res.ok) throw new Error(`Magento ${res.status}`);
  return res.json();
}

async function ordersSortedByIncrement(pageSize = 200, currentPage = 1) {
  const params = {
    "searchCriteria[sortOrders][0][field]": "increment_id",
    "searchCriteria[sortOrders][0][direction]": "ASC",
    "searchCriteria[pageSize]": pageSize,
    "searchCriteria[currentPage]": currentPage,
  };
  const data = await magentoGet("/orders", params);
  return data.items;
}

async function* allOrdersSortedByIncrement(pageSize = 200) {
  let page = 1;
  while (true) {
    const items = await ordersSortedByIncrement(pageSize, page);
    if (!items.length) return;
    for (const item of items) yield item;
    if (items.length < pageSize) return;
    page += 1;
  }
}

function normalizeOrder(item) {
  return {
    entityId: item.entity_id,
    storeId: item.store_id,
    incrementId: item.increment_id,
    createdAt: item.created_at,
  };
}

function parsePrefixByStore(raw) {
  const map = {};
  for (const pair of (raw || "").split(",")) {
    if (pair.includes(":")) {
      const [storeId, prefix] = pair.split(":");
      map[Number(storeId.trim())] = prefix.trim();
    }
  }
  return map;
}

export async function run() {
  const rawOrders = [];
  for await (const item of allOrdersSortedByIncrement(PAGE_SIZE)) rawOrders.push(item);
  const orders = rawOrders.map(normalizeOrder);

  const prefixByStore = parsePrefixByStore(process.env.PREFIX_BY_STORE);
  const result = detectSequenceDrift(orders, prefixByStore, GAP_THRESHOLD);

  for (const d of result.duplicates) {
    console.warn(`Store ${d.storeId}: increment_id ${d.incrementId} is duplicated across entity_id ${JSON.stringify(d.entityIds)}.`);
  }
  for (const g of result.gaps) {
    console.warn(`Store ${g.storeId}: gap of ${g.gapSize} between increment ${g.fromIncrement} and ${g.toIncrement}.`);
  }

  const affectedStores = new Set([
    ...result.duplicates.map((d) => d.storeId),
    ...result.gaps.map((g) => g.storeId),
  ]);

  if (affectedStores.size) {
    for (const storeId of [...affectedStores].sort()) {
      const resetValue = (result.maxNumericByStore[storeId] || 0) + 1;
      console.error(
        `Store ${storeId} sequence drift detected. Recommended repair: ` +
        `ALTER TABLE sequence_order_${storeId} AUTO_INCREMENT = ${resetValue} (run by a DBA, not this script).`
      );
    }
    console.error(`${affectedStores.size} store(s) affected. Exiting non-zero. No sequence table was written.`);
    process.exit(1);
  } else {
    console.log(`Done. No sequence drift found across ${orders.length} order(s).`);
  }
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
